buildutils/src/ensure-max-old-space.ts

```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env node
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+/**
+ * This file is a temporary workaround to ensure that we can set
+ * max_old_space by default but allow others to override it.
+ *
+ * When node 10 support is dropped, we can remove this file as node 12
+ * should just grow as required.
+ *
+ * See: https://github.com/jupyterlab/jupyterlab/issues/7175
+ *
+ * @example
+ * node ensure-max-old-space.js real-cli.js arg1 arg2
+ */
+import { execFileSync } from 'child_process';
+
+const MAX_OLD_SPACE = '--max_old_space_size=4096';
+
+if (!process.env.NODE_OPTIONS) {
+  process.env.NODE_OPTIONS = MAX_OLD_SPACE;
+} else if (process.env.NODE_OPTIONS.indexOf('max_old_space_size') === -1) {
```

Comment:
Slightly more robust:
```suggestion
} else if (process.env.NODE_OPTIONS.toLowerCase().indexOf('max_old_space_size') === -1) {
```

Reply:
Have you tested that non-lowercase is valid?

```
$ cat foo.js 
const v8 = require('v8');
console.log(v8.getHeapSpaceStatistics())
$ NODE_OPTIONS=--max_old_space_size=4096 node foo.js 
[ { space_name: 'read_only_space',
    space_size: 524288,
    space_used_size: 35200,
    space_available_size: 480384,
    physical_space_size: 43904 },
  { space_name: 'new_space',
    space_size: 2097152,
    space_used_size: 807640,
    space_available_size: 223528,
    physical_space_size: 2066192 },
  { space_name: 'old_space',
    space_size: 2330624,
    space_used_size: 2133688,
    space_available_size: 0,
    physical_space_size: 2177208 },
  { space_name: 'code_space',
    space_size: 1048576,
    space_used_size: 556544,
    space_available_size: 0,
    physical_space_size: 589312 },
  { space_name: 'map_space',
    space_size: 536576,
    space_used_size: 331320,
    space_available_size: 0,
    physical_space_size: 348728 },
  { space_name: 'large_object_space',
    space_size: 0,
    space_used_size: 0,
    space_available_size: 4394851840,
    physical_space_size: 0 } ]
$ NODE_OPTIONS=--MAX_OLD_SPACE_SIZE=10000 node foo.js 
node: --MAX_OLD_SPACE_SIZE= is not allowed in NODE_OPTIONS
```

Reply:
oh, also IIRC, dashes work in the environment: `--max-old-space-size` - we should probably check that as well.

Reply:
> Have you tested that non-lowercase is valid?

The question is what our command will do if the user adds an all caps option. Given the input
`NODE_OPTIONS=--MAX_OLD_SPACE_SIZE=10000`, we will change it to `NODE_OPTIONS=--MAX_OLD_SPACE_SIZE=10000 --max_old_space_size=4096`. I'm not sure what will happen then, but I don't think that is good. By calling `toLowerCase()`, we will pass the broken version through, and it will fail (as it should).

Reply:
> oh, also IIRC, dashes work in the environment: --max-old-space-size - we should probably check that as well.

Surprisingly  `--max_old_space-size` does as well! I have moved this to a regexp

> The question is what our command will do if the user adds an all caps option. Given the input
NODE_OPTIONS=--MAX_OLD_SPACE_SIZE=10000

They can pass many wrong options and we'd still have problems. `MAX_OLD_SPACE_SIZE` != `max_old_space_size` so we should let NODE take care of this. And it does:
```
NODE_OPTIONS=--MAX_OLD_SPACE_SIZE=1000 ./buildutils/lib/ensure-max-old-space.js node foo.js 
node: --MAX_OLD_SPACE_SIZE= is not allowed in NODE_OPTIONS
```
In the above code, foo.js would have got `NODE_OPTIONS="--MAX_OLD_SPACE_SIZE=1000 --max_old_space_size=4096"`

Reply:
@vidartf @jasongrout Any reason to not merge this still?